rlbot_gui/story/story_challenge_setup.py

```diff
@@ -0,0 +1,462 @@
+            new_score = gamePacket.teams[i].score
+            if new_score != self._last_score_by_team[team_index]:
+                self._last_score_by_team[team_index] = new_score
+
+                if team_index == self._human_team:
+                    last_touch_player = self._last_touch_by_team[
+                        team_index
+                    ].player_index
+                    if last_touch_player == self._human_player_index:
+                        self.stats["humanGoalsScored"] += 1
+                        print("humanGoalsScored")
+
+
+def manage_game_state(
+    challenge: dict, upgrades: dict, setup_manager: SetupManager
+) -> Tuple[bool, dict]:
+    """
+    Continuously track the game and adjust state to respect challenge rules and
+    upgrades.
+    At the end of the game, calculate results and the challenge completion
+    and return that
+    """
+    early_failure = False, {}
+
+    expected_player_count = challenge["humanTeamSize"] + len(challenge["opponentBots"])
+    # Wait for everything to be initialized
+    packet = GameTickPacket()
+    setup_manager.game_interface.fresh_live_data_packet(packet, 1000, WITNESS_ID)
+    waiting_start = time.monotonic()
+    while packet.num_cars != expected_player_count and time.monotonic() - waiting_start < 5:
+        print("Game started but no cars are in the packets")
+        time.sleep(0.5)
+        setup_manager.game_interface.fresh_live_data_packet(packet, 1000, WITNESS_ID)
+
+    if packet.num_cars == 0:
+        print("The game was initialized with no cars")
+        return early_failure
+
+    tick_rate = 120
+    results = None
+    max_boost = 0
+    if "boost-100" in upgrades:
+        max_boost = 100
+    elif "boost-33" in upgrades:
+        max_boost = 33
+
+    half_field = challenge.get("limitations", []).count("half-field") > 0
+
+    stats_tracker = ManualStatsTracker(challenge)
+    last_boost_bump_time = time.monotonic()
+    while True:
+        try:
+            packet = GameTickPacket()
+            setup_manager.game_interface.fresh_live_data_packet(
+                packet, 1000, WITNESS_ID
+            )
+
+            if packet.num_cars == 0:
+                # User seems to have ended the match
+                print("User ended the match")
+                return early_failure
+
+            stats_tracker.updateStats(packet)
+            results = packet_to_game_results(packet)
+
+            if has_user_perma_failed(challenge, stats_tracker.stats):
+                time.sleep(3)
+                setup_failure_freeplay(setup_manager, "You failed the challenge!")
+                return early_failure
+            
+            if end_by_mercy(challenge, stats_tracker.stats, results):
+                time.sleep(5)
+                setup_failure_freeplay(setup_manager, "Challenge completed by mercy rule!", "green")
+                return True, results
+
+            game_state = GameState.create_from_gametickpacket(packet)
```

Comment:
This is slightly sub-optimal. It creates a game state with everything filled in, including position data that might be 1 frame stale if the user is unlucky / has a low performing system. This can result in tiny amounts of rubber banding, and in rare cases weird ball hits. If you leave a part of the game state blank, we do a better job of leaving it undisturbed.

Reply:
Oh wow, I didn't realize that's how it worked! Thanks for the insight. It does seem to perform better now.